File: issn.ts
export const ISSNPattern: string = "^ISSN\s+?(\d{4})-(\d{3}[\dX])$";
export const reISSN: RegExp = new RegExp(ISSNPattern, "i");

/**
 * Validates the checksum of an ISSN.
 * @param issn - The normalized ISSN string (without hyphen).
 * @returns boolean - True if valid, otherwise false.
 */
function validateISSNChecksum(issn: string): boolean {
  if (issn.length !== 8) return false;

  const digits = issn.slice(0, 7);
  const checkDigit = issn[7].toUpperCase();

  const checksum = digits
    .split("")
    .reduce((sum, char, index) => sum + parseInt(char) * (8 - index), 0) %
    11;

  const expectedCheckDigit = checksum === 0
    ? "0"
    : checksum === 1
    ? "X"
    : `${11 - checksum}`;
  return checkDigit === expectedCheckDigit;
}

function stripISSN(issn: string): string {
  return issn.toUpperCase().replaceAll(/\D|X/g, "");
}

/**
 * Validates the format and checksum of an ISSN.
 * @param issn - The ISSN string to validate.
 * @returns boolean - True if the ISSN is valid, otherwise false.
 */
export function validateISSN(issn: string): boolean {
  const bareISSN = stripISSN(issn);
  return validateISSNChecksum(bareISSN);
}

/**
 * Normalizes an ISSN by removing spaces, invalid characters, and ensuring uppercase.
 * @param issn - The ISSN string to normalize.
 * @returns string - The normalized ISSN string (without hyphen).
 */
export function normalizeISSN(issn: string): string {
  const bareISSN = stripISSN(issn);
  //console.log(`DEBUG bareISSN -> ${bareISSN}`);
  return `${bareISSN.substring(0, 4)}-${bareISSN.substring(4)}`;
}

File: issn.go
package metadatatools

import "strconv"

// stripISSN removes any non-digit characters from issn.
func stripISSN(issn string) string {
	return reNonDigit.ReplaceAllString(issn, "")
}

// validateISSNChecksum validates the Mod 11 checksum digit of a
// normalized, 8 digit ISSN.
func validateISSNChecksum(issn string) bool {
	if len(issn) != 8 {
		return false
	}

	digits := digitsOf(issn[:7])
	sum := 0
	for i, d := range digits {
		sum += d * (8 - i)
	}
	checksum := sum % 11

	var expected string
	switch checksum {
	case 0:
		expected = "0"
	case 1:
		expected = "X"
	default:
		expected = strconv.Itoa(11 - checksum)
	}
	return issn[7:8] == expected
}

// ValidateISSN normalizes issn then validates its checksum digit.
func ValidateISSN(issn string) bool {
	return validateISSNChecksum(stripISSN(issn))
}

// NormalizeISSN removes any non-digit characters from issn and
// reformats it as NNNN-NNNN.
//
// Example:
//
//	issn := NormalizeISSN("1058-6180")
//	// issn == "1058-6180"
func NormalizeISSN(issn string) string {
	return joinGroups(stripISSN(issn), "-", 4)
}
